modules/test-runner/src/main/scala/scala/build/testrunner/FrameworkUtils.scala

```diff
@@ -0,0 +1,223 @@
+    val distinctFrameworks = distinctBy(allFrameworks)(_.name())
+    if (distinctFrameworks.nonEmpty)
+      logger.debug(
+        s"""Distinct test frameworks found (by framework name):
+           |  - ${distinctFrameworks.map(_.description).mkString("\n  - ")}
+           |""".stripMargin
+      )
+
+    val finalFrameworks =
+      distinctFrameworks
+        .filter(f1 =>
+          !distinctFrameworks
+            .filter(_ != f1)
+            .exists(f2 =>
+              f1.getClass.isAssignableFrom(f2.getClass)
+            )
+        )
+    if (finalFrameworks.nonEmpty)
+      logger.log(
+        s"""Final list of test frameworks found:
+           |  - ${finalFrameworks.map(_.description).mkString("\n  - ")}
+           |""".stripMargin
+      )
+
+    val skippedInheritedFrameworks = distinctFrameworks.diff(finalFrameworks)
+    if (skippedInheritedFrameworks.nonEmpty)
+      logger.log(
+        s"""The following test frameworks have been filtered out, as they're being inherited from by others:
+           |  - ${skippedInheritedFrameworks.map(_.description).mkString("\n  - ")}
+           |""".stripMargin
+      )
+
+    finalFrameworks
+  }
+
+  def getFrameworksToRun(
+    frameworkServices: Seq[Framework],
+    frameworks: Seq[Framework]
+  )(logger: Logger): Seq[Framework] = {
+    if (frameworkServices.nonEmpty)
+      logger.debug(
+        s"""Found test framework services:
+           |  - ${frameworkServices.map(_.description).mkString("\n  - ")}
+           |""".stripMargin
+      )
+    if (frameworks.nonEmpty)
+      logger.debug(
+        s"""Found test frameworks:
+           |  - ${frameworks.map(_.description).mkString("\n  - ")}
+           |""".stripMargin
+      )
+
+    getFrameworksToRun(allFrameworks = frameworkServices ++ frameworks)(logger)
+  }
+
+  def listClasses(classPath: Seq[Path], keepJars: Boolean): Iterator[String] =
+    classPath.iterator.flatMap(listClasses(_, keepJars))
+
+  def listClasses(classPathEntry: Path, keepJars: Boolean): Iterator[String] =
+    if (Files.isDirectory(classPathEntry)) {
+      var stream: java.util.stream.Stream[Path] = null
+      try {
+        stream = Files.walk(classPathEntry, Int.MaxValue)
+        stream
+          .iterator
+          .asScala
+          .filter(_.getFileName.toString.endsWith(".class"))
+          .map(classPathEntry.relativize)
+          .map { p =>
+            val count = p.getNameCount
+            (0 until count).map(p.getName).mkString(".")
+          }
+          .map(_.stripSuffix(".class"))
+          .toVector // fully consume stream before closing it
+          .iterator
+      }
+      finally if (stream != null) stream.close()
+    }
+    else if (keepJars && Files.isRegularFile(classPathEntry)) {
+      import java.util.zip._
+      var zf: ZipFile = null
+      try {
+        zf = new ZipFile(classPathEntry.toFile)
+        zf.entries
+          .asScala
+          // FIXME Check if these are files too
```

Comment:
Should we just add a .exists check here? Might be worth updating or resolving this FIXME

Reply:
I can try to look into it, but it's out of scope for this PR.
I merely moved this piece of code around.